### ISSUE TITLE
[BUG] AttachVolume.Attach failed for volume: volume is not ready for workloads 

### ISSUE THREAD

**Issue by CC007** [longhorn/longhorn#9511]
## Describe the bug
I get the following error when I try to deploy a postgresql database to my cluster:
```
AttachVolume.Attach failed for volume "pvc-20544069-75fd-4103-829e-d3091e5b6ced" : rpc error: code = Aborted desc = volume pvc-20544069-75fd-4103-829e-d3091e5b6ced is not ready for workloads 
```
This prevents the pod from starting.

It looks like the pvc and pv are bound correctly

## To Reproduce
 - Install longhorn (This worked successfully and I was able to create volumes for Grafana and Nexus using longhorn as storageClass)
 - Install the postgres operator
```bash
# add repo for postgres-operator
helm repo add postgres-operator-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator

# install the postgres-operator
helm install postgres-operator postgres-operator-charts/postgres-operator
```
 - (optionally) Install the postgres operator UI
```bash
# add repo for postgres-operator-ui
helm repo add postgres-operator-ui-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator-ui

# install the postgres-operator-ui
helm install postgres-operator-ui postgres-operator-ui-charts/postgres-operator-ui
```
 - create a postgresql resource (making sure to use the storageClass "longhorn"). This yaml file was created using the postgres operator UI.:
   
 ```yaml
 kind: "postgresql"
apiVersion: "acid.zalan.do/v1"

metadata:
  name: "charlotte-db"
  namespace: "database"
  labels:
    team: xenatecc

spec:
  teamId: "xenatecc"
  postgresql:
    version: "16"
  numberOfInstances: 1
  maintenanceWindows:
  volume:
    size: "16Gi"
    storageClass: "longhorn"
  users:
    xenatecc: []
  databases:
    charlotte: xenatecc
  allowedSourceRanges:
  # IP ranges to access your cluster go here

  resources:
    requests:
      cpu: 100m
      memory: 100Mi
    limits:
      cpu: 500m
      memory: 500Mi
```
   

## Expected behavior

The resources (statefulset, pod, pv, pvc, etc.) get created, pv and pvc get bound and the pod succesfully started.

## Environment

 - Longhorn version: 1.6.2
 - Impacted volume (PV): pvc-20544069-75fd-4103-829e-d3091e5b6ced 
   - For PVC: pgdata-charlotte-db-0 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: K3s v1.30.3+k3s1 
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 5
 - Node config
   - Hardware: 8x Raspberry Pi 4B (arm64)
   - OS type and version: Ubuntu 22.04.3 LTS
   - Kernel version: Linux 5.15.0-1061-raspi
   - CPU per node: 4
   - Memory per node: 8GB
   - Disk type (e.g. SSD/NVMe/HDD): 4x NVMe SSD, connected to 4 Raspberry Pis using iSCSI (see image below)
   - Network bandwidth between the nodes (Gbps): 1GB
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): baremetal
 - Number of Longhorn volumes in the cluster: 3 (2 working, 1 not working)

## Additional context
Logs:
[longhorn-manager-4l89v_longhorn-manager.log](https://github.com/user-attachments/files/17085929/longhorn-manager-4l89v_longhorn-manager.log)

Images
![afbeelding](https://github.com/user-attachments/assets/faaeb018-fbc4-4291-9a0d-70ca1b819447)
![afbeelding](https://github.com/user-attachments/assets/521614de-4ae4-4883-9913-c0efe004e873)

![afbeelding](https://github.com/user-attachments/assets/fdb729e2-bbd1-4436-adb1-0ae170452bc8)

## Workaround and Mitigation
Couldn't find a way to work around it. I tried removing the volume, I tried deleting the postgres pod. Neither helped.

**Comment by CC007** [comment]
Support bundle was stuck at 40%, so I couldn't generate it. I tried again, but it complained that another support bundle was already being generated, but I can't find where I could find that progress.

**Comment by shuo-wu** [comment]
Found the below error in the longhorn-manager pod log 
```
time="2024-09-21T22:08:00Z" level=warning msg="pvc-20544069-75fd-4103-829e-d3091e5b6ced-e-0: time=\"2024-09-21T22:07:09Z\" level=warning msg=\"Failed to discover\" func=\"iscsidev.(*Device).StartInitator\" file=\"iscsi.go:161\" error=\"failed to execute: /usr/bin/nsenter [nsenter --mount=/host/proc/772/ns/mnt --net=/host/proc/772/ns/net iscsiadm -m discovery -t sendtargets -p 10.42.4.14], output , stderr iscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: cannot make connection to 10.42.4.14: No route to host\\niscsiadm: connection login retries (reopen_max) 5 exceeded\\niscsiadm: Could not perform SendTargets discovery: iSCSI PDU timed out\\n: exit status 11\"" func="controller.(*InstanceHandler).printInstanceLogs" file="instance_handler.go:467"
```

There is an existing KB article talked about this error, you can take a look first: https://longhorn.io/kb/troubleshooting-orphan-iscsi-session-error/

**Comment by CC007** [comment]
The IP listed in the logs does have an instance manager, so I don't think it's a orphan/stale iscsi session

**Comment by CC007** [comment]
Also

> "While annoying, the logs do not indicate any severe issues."

In my case, I do have issues, since the postgres pod doesn't successfully attach to the volume.

**Comment by CaptainKrby** [comment]
I've just emailed bundle support, I have exactly the same problem which occurred very recently on dozens of pods. 
My version of longhorn 1.6.0 has been very stable over the months until now. It runs in a Kubernetes 1.28.10 cluster on Bare Metal.

**Comment by CC007** [comment]
I found out why the support bundle creation is failing:

[supportbundle-agent-support-bundle-2024-09-21t22-44-09z-4gkc5_agent.log](https://github.com/user-attachments/files/17175987/supportbundle-agent-support-bundle-2024-09-21t22-44-09z-4gkc5_agent.log)

```
...

+ curl -v -i -H 'Content-Type: application/zip' --data-binary @node_bundle.zip http://10.42.2.25:8080/nodes/pirack1-node3
curl: option --data-binary: out of memory
curl: try 'curl --help' or 'curl --manual' for more information
```

**Comment by CC007** [comment]
I've did a little experiment.

I tried creating a PVC and a deployment:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pv-poc
spec:
  accessModes:
    - ReadWriteOnce
  volumeMode: Filesystem
  resources:
    requests:
      storage: 8Gi
  storageClassName: longhorn
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: pv-poc
  labels:
    app: pv-poc
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pv-poc
  template:
    metadata:
      name: pv-poc
      labels:
        app: pv-poc
    spec:
      containers:
        - name: pv-poc
          image: nginx
          imagePullPolicy: IfNotPresent
          volumeMounts:
            - mountPath: /data/pv-poc/mnt
              name: pv-poc
          ports:
            - containerPort: 80
              protocol: TCP
      restartPolicy: Always
      volumes:
        - name: pv-poc
          persistentVolumeClaim:
            claimName: pv-poc
  ```
  This seems to bind to a longhorn PV fine and the volume attaches properly.
  I then tried with a StatefulSet:
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: pv-poc
  labels:
    app: pv-poc
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pv-poc
  serviceName: pv-poc
  template:
    metadata:
      name: pv-poc
      labels:
        app: pv-poc
    spec:
      containers:
        - name: pv-poc
          image: nginx
          imagePullPolicy: IfNotPresent
          volumeMounts:
            - mountPath: /data/pv-poc/mnt
              name: pv-poc
          ports:
            - containerPort: 80
              protocol: TCP
  volumeClaimTemplates:
    - metadata:
        name: pv-poc
      spec:
        accessModes:
         - ReadWriteOnce
        storageClassName: longhorn
        resources:
          requests:
            storage: 8Gi
```
Here I ran into a similar issue as above: 
```
AttachVolume.Attach failed for volume "pvc-6f8daadc-7cbe-4232-b10c-15d1c0074bf7" : rpc error: code = DeadlineExceeded desc = volume pvc-6f8daadc-7cbe-4232-b10c-15d1c0074bf7 failed to attach to node pirack2-node3 with attachmentID csi-fb351d9a670fe0ef7225a85402d667da9f972911ae16fd0b7eebed5868dbc642 
```
I found this other issue that might be related: https://github.com/longhorn/longhorn/issues/5990

**Comment by CC007** [comment]
Correction: the PVC + deployment now also doesn't seem to work consistently anymore.

**Comment by CaptainKrby** [comment]
It's even worse when I restart a node, all pods that rely on Longhorn remain stuck on ContainerCreating state with the same message: 

`AttachVolume.Attach failed for volume “pvc-9e9eee9c-06fe-4541-b8aa-3d9accb1a097” : rpc error: code = Aborted desc = volume pvc-9e9eee9c-06fe-4541-b8aa-3d9accb1a097 is not ready for workloads`

**Comment by CC007** [comment]
I updated to v1.7.1 and did a staggered `apt update` + `apt upgrade` + `reboot` of all my nodes. Now all the above issues seem to be resolved, for the pv-poc pvc+deployment, pv-poc statefulset as well as the postgres deployment.

I don't know which of the above fixed the issue (or if it was a fluke), but it's fixed for me now. 

@CaptainKrby If you still experience the problem on 1.7.1+, I think it would be best if you create your own issue.

**Comment by sirajkrm** [comment]
for anyone facing the issue, in my case I had to reboot the node a couple of times to resolve this.

the longhorn-manager pod was showing an error message like this
```
level=warning msg="Failed to check endpoint https://longhorn-conversion-webhook.longhorn-system.svc:9501/v1/healthz" func=webhook.isServiceAvailable file="webhook.go:78" error="Get \"https://longhorn-conversion-webhook.longhorn-system.svc:9501/v1/healthz\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
``` 

decided to check the networking side by creating 2 simple pods with ubuntu image that contain curl command
1 pod on a healthy node that gave 200 OK http code which means DNS resolution is okay and the service is reachable
1 pod on a faulty node that I wasn't even able to ssh into and exec the curl command

the faulty node was also having a high resource consumption so I did a (first) reboot

the longhorn-manager then shows this error message
```
Successfully assigned dvp-ns/rabbitmq-0 to k8s-node-12
  Warning  FailedAttachVolume  55s   attachdetach-controller  AttachVolume.Attach failed for volume "pvc-710329a0-xxxxxx-yyyyyy-zzzzzzzzz" : rpc error: code = Aborted desc = volume pvc-710329a0-xxxxxx-yyyyyy-zzzzzzzzz is not ready for workloads
```

by chance I checked the firewall and it was active so I stopped it and disabled it.

I made a second reboot of the node and right after that the longhorn-manager got re-created successfully and run just fine.
Longhorn took care of the faulty volumes already.
I was scaling down the workload (mostly statefulsets and couple of deployments) and slowly scaling up made it to a running state and volumes got attached with success.
